### PR TITLE
Login

### PR DESCRIPTION
**Pull request by GrooveStomp**
There are a ton of changes here adding some pretty massive functionality.
- Session handling via psql.
- Login + Logout.
- Partial templates and logic in templates.

### CODE REVIEW

File: types.go [hunk 1]
```diff
@@ -22,7 +23,15 @@ type HttpServerConfig struct {
...
 type User struct {
-	Id       sql.NullString
-	Username string
-	Password string
+	Id        sql.NullString
+	Username  string
+	Password  string
+	CreatedAt time.Time      `db:"created_at"`
+	UpdatedAt time.Time      `db:"updated_at"`
+	DeletedAt sql.NullString `db:"deleted_at"`
```
Comment (GrooveStomp) [inline types.go:31]: These serialization comments are probably not required.

**Comment by GrooveStomp** [comment]
[Logout](https://trello.com/c/Z6NTiDjb/3-logout)

**Comment by GrooveStomp** [comment]
[Login](https://trello.com/c/MbP4ORqH/2-login)

**Comment by GrooveStomp** [comment]
[Use go modules from Go 1.11 to lock dependencies and create reproducible builds.](https://trello.com/c/zav5EVpq/15-use-go-modules-from-go-111-to-lock-dependencies-and-create-reproducible-builds)

File: types.go (hunk 1 above)
Comment (GrooveStomp) [reply types.go:31]: On further investigation, these do, in fact, apear to be necessary.

File: request_handlers.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+func CreateUser(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	logHandlerIntro(r.Method, r.URL.Path, r.Form)
+
+	newUser := User{
+		Username: r.Form["username"][0],
+		Password: r.Form["password"][0],
+	}
+
+	row := db.QueryRowx(
```
Comment (Trinergy) [inline request_handlers.go:37]: I assume we don't care about security at the moment? (SQL Injection delicious) @GrooveStomp

File: request_handlers.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+		row := db.QueryRowx(`INSERT INTO authentications (user_id) VALUES ($1) RETURNING token`, userId)
+		err = row.Scan(&authToken)
+		if err != nil {
+			log.Println(err)
+			http.Error(w, "Error logging in", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Invalidate all outstanding authentications for this user.
```
Comment (Trinergy) [inline request_handlers.go:101]: (Noob Question) Is this how flushing authentication tokens normally works? Wouldn't you eventually accumulate a lot of `deleted = true` authentication tokens and make queries slower? Or do most apps do this to keep track of all login events of a user?

File: request_handlers.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+	log.Printf("findLoginAuthToken: authToken: %s\n", authToken)
+	return authToken, nil
+}
+
+func renderPage(w http.ResponseWriter, r *http.Request, templateName, title string) {
+	userId, _ := authenticate(w, r)
+
+	t := template.Must(
+		template.ParseFiles(
```
Comment (Trinergy) [inline request_handlers.go:196]: this is pretty damn cool 🍇

File: templates/index.tmpl [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{{ define "home" }}
+{{ template "header" . }}
```
Comment (Trinergy) [inline templates/index.tmpl:2]: What's the `.` ? just path of file for the defined variable? Why is that needed if you already defined it in `template.ParseFiles`?

File: templates/navigation.tmpl [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{{ define "navigation" }}
+<ul>
+  {{ if ne .Title "home" }}<li><a href="/">Home</a></li>{{ end }}
```
Comment (Trinergy) [inline templates/navigation.tmpl:3]: What is `ne` supposed to be?

**Review by Trinergy — APPROVED** [review]
Left some questions. Such readable code, much amaze.

File: request_handlers.go (hunk 3 above)
Comment (GrooveStomp) [reply request_handlers.go:101]: I don't know.
I figure leaving authentications in for audit trails might be a worthwhile thing.
I *do* need to add appropriate indices in, though.

File: templates/index.tmpl (hunk 5 above)
Comment (GrooveStomp) [reply templates/index.tmpl:2]: `.` Refers to the object being passed into the template.
At the caller I'm passing in the struct, but within the template itself, the struct passed in is referred to as `.`.
Any partials being rendered by the invoked template don't receive the initial object passed in, so this is how you accomplish that.

I'm kind of stumbling over my own words there.
If you look at the first example shown in the overview here: https://golang.org/pkg/text/template/
See how the attribute is accessed via `.Material`?  We don't have a name for the object being passed in... I think this way of just invoking the attribute on nothing (ie., `.Material`) or referring to the currently active object via `.` is just special syntax exclusive to templates in Golang.  I'm honestly not too sure.

File: templates/navigation.tmpl (hunk 6 above)
Comment (GrooveStomp) [reply templates/navigation.tmpl:3]: Not Equal. See: https://golang.org/pkg/text/template/#hdr-Functions

File: request_handlers.go (hunk 2 above)
Comment (GrooveStomp) [reply request_handlers.go:37]: **sqlx**'s APIs protect against SQL injection, see: https://jmoiron.github.io/sqlx/#bindvars
Bindvars for PostgreSQL are in the form `$1`, `$2`, etc.

**Comment by GrooveStomp** [comment]
> https://github.com/Go-Team-Gamma/classroom-seating/pull/7#pullrequestreview-178100851

Thanks.
I'm still coming to terms with Golang. I wouldn't say I think it's the most easily readable, but I am learning as I go and I think you can see some improvements even moving from this branch to the unit testing branch.
